akka-projection-grpc/src/main/scala/akka/persistence/EventWriter.scala

```diff
@@ -0,0 +1,339 @@
+                    handleUpdatedStateForPid(pid, newState)
+                  } else {
+                    // ack all already written
+                    val stateAfterWritten = alreadyInJournal.foldLeft(state) { (state, seqNr) =>
+                      val (_, replyTo) = state.waitingForReply(seqNr)
+                      replyTo ! StatusReply.success(WriteAck(pid, seqNr))
+                      state.copy(waitingForReply = state.waitingForReply - seqNr)
+                    }
+                    if (needsWrite.isEmpty) {
+                      handleUpdatedStateForPid(pid, stateAfterWritten)
+                    } else {
+                      // retrigger write for those left if any
+                      val reprsToRewrite =
+                        stateAfterWritten.waitingForReply.values.map { case (repr, _) => repr }.toVector
+                      if (context.log.isDebugEnabled())
+                        context.log.debugN(
+                          "Partial batch was duplicates, re-triggering write of persistence id [{}], sequence nr [{}-{}]",
+                          pid,
+                          reprsToRewrite.head.sequenceNr,
+                          reprsToRewrite.last.sequenceNr)
+                      val write = AtomicWrite(reprsToRewrite) :: Nil
+                      journal ! JournalProtocol.WriteMessages(write, context.self.toClassic, actorInstanceId)
+                    }
+
+                  }
+
+              }
+              Behaviors.same
+
+            case response: JournalProtocol.Response => handleJournalResponse(response)
+
+            case unexpected =>
+              context.log.warn("Unexpected message sent to EventWriter [{}], ignored", unexpected.getClass)
+              Behaviors.same
+
+          }
+        })
+      .onFailure[Exception](SupervisorStrategy.restart)
+      .narrow[Command]
+
+  object MaxSeqNrFinder {
+    def apply(
+        replyTo: ActorRef[EventWriter.MaxSeqNrForPid],
+        journal: akka.actor.ActorRef,
+        persistenceId: String,
+        fromSeqNr: Long,
+        toSeqNr: Long,
+        originalErrorDesc: String): Behavior[Nothing] =
+      Behaviors
+        .setup[AnyRef] { context =>
+          journal ! JournalProtocol.ReplayMessages(
+            fromSeqNr,
+            toSeqNr,
+            Long.MaxValue,
+            persistenceId,
+            context.self.toClassic)
+
+          waitingForReplay(replyTo, persistenceId, 0L, originalErrorDesc)
+        }
+        .narrow
+
+    def waitingForReplay(
+        replyTo: ActorRef[EventWriter.MaxSeqNrForPid],
+        persistenceId: String,
+        maxSequenceNumber: Long,
+        originalErrorDesc: String): Behavior[AnyRef] = {
+
+      Behaviors.receiveMessage {
+        case JournalProtocol.ReplayedMessage(repr) =>
+          waitingForReplay(replyTo, persistenceId, repr.sequenceNr, originalErrorDesc)
+        case JournalProtocol.RecoverySuccess =>
```

Comment:
isn't this wrong? shouldn't it be
```
case RecoverySuccess(highestSequenceNr: Long)
```

Reply:
Yes, and we don't need to keep track of highest from replayed event either, changed in 668e4dbf0904e7d8e1aad17e9aee210a9088bf9d